Prevent accidental disclosure of sensitive files during development

tahb opened the pull request:
Based atop https://github.com/dxw/rails-template/pull/213

## Changes in this PR

- preventing accidental disclosure
- refactor Dockerfile into multi-stage builds. This allows us to `COPY` all the extra files needed for testing into an isolated stage. This allows us to separate copied files needed for the app in prod from those needed only for test

## Preventing accidental disclosure

As an action from an incident on DfE Schools Buying we would like to push this change upstream.

Using `COPY . .` is problematic as it could (and has) quietly included sensitive files such as `.env` which was then pushed to a Docker registry. This allows anyone who pulls that image to read the contents are get hold of secrets etc.

Instead of doing this we can change to a whitelist approach. Including only the folders we expect for a Rails project. These should remain static and perhaps only change during major Rails releases.

A `.dockerignore` file does already exist as a mechanism to achieve the same result however it requires a human to rememeber to configure it properly -  steps have also been made to improve this file by default [1]. This Dockerfile change is therefore aimed at provided more defence in depth to help prevent accidental disclosure of sensitive files.

We are using this in production already at least on DfE Schools Buying [2] and DfE Claim [3].


[1] https://github.com/dxw/rails-template/pull/198
[2] https://github.com/DFE-Digital/buy-for-your-school/blob/develop/Dockerfile#L33
[3] https://github.com/DFE-Digital/claim-additional-payments-for-teaching/blob/master/Dockerfile#L101


## Next steps

Move the incident action card to done https://trello.com/c/A8DBTFbM

tahb committed:
Prevent accidental disclosure of sensitive files during development

Using `COPY . .` is problematic as it could (and has) quietly included sensitive files such as `.env` which was then pushed to a Docker registry.

Instead of doing this we can change to a whitelist approach. Including only the  folders we expect for a Rails project. These should remain static and perhaps only change during major Rails releases.

A `.dockerignore` file does already exist as a mechanism to achieve the same result however it requires a human to rememeber to configure it properly -  steps have also been made to improve this file by default [1]. This Dockerfile change is therefore aimed at provided more defence in depth to help prevent accidental disclosure of sensitive files.

We are using this in production already at least on DfE Schools Buying [2] and DfE Claim [3].

We have to copy in the dotfiles for our linters so they're available. Next step is to use docker layers to isolate these dependencies from the final build image that's run in prod.

[1] https://github.com/dxw/rails-template/pull/198
[2] https://github.com/DFE-Digital/buy-for-your-school/blob/develop/Dockerfile#L33
[3] https://github.com/DFE-Digital/claim-additional-payments-for-teaching/blob/master/Dockerfile#L101

rjw1 reviewed (commented):
Did you mean for the WIP commit for package.json stuff to make it into this PR?

tahb commented:
@rjw1 Yes and no! CI wasn't passing on Friday so I tried to see if I could fix it quickly before spinning off another branch so figure out why testing this app doesn't work properly. I'll rebase that WIP commit away 👍 but this PR is otherwise not ready.

tahb committed:
Dockerfile uses multi-stage builds

This change provides several benefits:

* We can avoid doing `COPY . .` in the Dockerfile _without_ having to manually copy all the files needed for running the tests
* Files used solely for testing the application are not included in the Docker image by default. When deployments are enabled we will need to set a `--target` when building for deployment. If we don't set a `--target` stage the last `test` stage will be used, which will only run the specs. This should be a safe prompt for us to remember to provide the `web` target.
* Smaller docker images
* Inspired by our latest RODA Dockerfile [1]

[1] https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/Dockerfile

Floppy commented:
@tahb can you look at resolving the Dockerfile conflict, so we can get this approved and merged in?

tahb commented:
@Floppy Thanks for reviewing 🙌 I'll find sometime Friday to get to this. The other PR https://github.com/dxw/rails-template/pull/213 will need to be merged first and this will need a rebase anyway but it can't hurt to keep it up to date if there's time.

Floppy commented:
Ah, cool, let's nail some people down to get #213 merged first then.

tahb commented on Dockerfile:
```diff
@@ -81,4 +103,22 @@ ENTRYPOINT ["/docker-entrypoint.sh"]
 EXPOSE 3000
 
-CMD ["rails", "server"]
+CMD ["bundle", "exec", "rails", "server"]
```
We shouldn't _need_ to write `bundle exec` to ensure we are always using the expected versions of gems. It doesn't hurt to add this here to protect against any breaking change to environment variables. We are asking bundler to load first, and to figure out how we want to require the rails dependency.

tahb commented:
@rjw1 @Floppy The dependency PR for this has been merged. We are now good to go if you're still able to cast an eye across these 2 commits. 

@leeky @dragon-dxw if you're available for reviewing this too that'd be great. I'm trying to get the template up to date before we use it for CitA next week!

dragon-dxw commented on Dockerfile:
```diff
@@ -8,35 +11,79 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
 ARG RAILS_ENV
-ENV RAILS_ENV=${RAILS_ENV:-production}
-ENV RACK_ENV=${RAILS_ENV:-production}
+ENV RAILS_ENV ${RAILS_ENV:-production}
```
Is there a semantic difference between `ENV FOO=bar` and `ENV FOO bar` ?
If we're standardising there's also BUNDLE_GEM_GROUPS, CURRENT_SHA, TIME_OF_BUILD which might want the same treatment.
[not blocking]

dragon-dxw replied:
Found the answer -- https://docs.docker.com/engine/reference/builder/#env :

There's subtle issues with formatting (ENV A=X B=Y sets two enviroment variables to X and Y as you'd expect, A X B=Y sets A to `X B=Y`); spaces may be deprecated in future.
There's perhaps an argument to be made that ENV A="X" is more correct but this isn't a priority.

dragon-dxw reviewed (approved):
Couldn't run this branch or develop in my environment -- see https://dxw.slack.com/archives/C018M33K3SB/p1632845151014300 but that's no reason to block this. 
Looks like it makes things significantly nicer.

dragon-dxw commented on Dockerfile:
```diff
@@ -8,35 +11,79 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
+COPY config ${APP_HOME}/config
+COPY db ${APP_HOME}/db
+COPY script ${APP_HOME}/script
+COPY app ${APP_HOME}/app
```
There's generally a thing in Dockerfiles about big run-on commands to avoid having too many lines, each of which creates a different image (if that's the right nomenclature?)
I assume COPY is fine because we know it's not touching large files and making multiple copies of them, or something? (unlike the apt-get commands databases etc?)
[for my info only, non-blocking]

leeky replied:
From my brief digging, I believe each `COPY` adds an additional layer. However doing multiple `COPY`s together...

`COPY file.rb file.json directory/ another_directory/ ${APP_HOME}/`

...leads to the **contents** of `directory` and `another_directory` being added to `${APP_HOME}` rather than those folders themselves.

A quick search of Docker documentation and StackOverflow suggests the answer as being the one we're trying not to do `COPY . ${APP_HOME}/`.

@tahb Might be worth merging the root file copies into one line to cut down layers, maybe?

tahb replied:
Do you have an idea of the cost of having multiple docker layers? I know that layers are a part of how docker is able to cache builds which is especially useful for `app/`, given the contents will change more than all the rest. I assume that the performance difference is negligible and if it becomes a problem then we can make an improvement.

leeky reviewed (approved):
LGTM! :shipit: 

Looks like we're going to have to live with additional layers being created by the individual `COPY` commands as there doesn't seem to be away around it.

Additionally, FYI Node 10 is deprecated, and the build pauses for 20 seconds to display a warning. Obviously, this should only add 20 seconds the first time this layer is cached, but might be worth bumping to a newer Node version.

tahb commented:
@leeky thanks for highlighting the node upgrade. I've made a note on this PR to prompt us to deal with that https://github.com/dxw/rails-template/pull/288

Floppy reviewed (approved):
LGTM - well done everyone :)
